src/rgw/rgw_reshard.cc

```diff
@@ -1181,3 +1182,76 @@ std::ostream& RGWReshard::ReshardWorker::gen_prefix(std::ostream& out) const
 {
   return out << "rgw reshard worker thread: ";
 }
+
+int bilog_trim(const DoutPrefixProvider* p, rgw::sal::RGWRadosStore* store,
+	       RGWBucketInfo& bucket_info, uint64_t gen, int shard_id,
+	       std::string_view start_marker, std::string_view end_marker)
+{
+  auto& logs = bucket_info.layout.logs;
+  auto log = std::find_if(logs.begin(), logs.end(), rgw::matches_gen(gen));
+  if (log == logs.end()) {
+    ldpp_dout(p, 5) << __PRETTY_FUNCTION__ << ":" << __LINE__
+		    << "ERROR: no log layout with gen=" << gen << dendl;
+    return -ENOENT;
+  }
+
+  auto log_layout = *log; // current log layout for log listing
+
+  auto r = store->svc()->bilog_rados->log_trim(bucket_info, log_layout, shard_id,
+					       start_marker, end_marker);
+  if (r < 0) {
+    ldpp_dout(p, 5) << __PRETTY_FUNCTION__ << ":" << __LINE__
+		    << "ERROR: bilog_rados->log_trim returned r=" << r << dendl;
+    return r;
+  }
+
+  // Do not delete the only generation
+
+  if (logs.size() == 1) {
+    return 0;
+  }
+
+  // Check for emptiness
+
+  std::list<rgw_bi_log_entry> l;
```

Comment:
I assume std::list has always been used for this

Reply:
Yes.